File: scala/109.scala
object Solution {
    def sortedListToBST(head: ListNode): TreeNode = {
        if (head == null) null
        else if (head.next == null) TreeNode(head.x)
        else {
          var slow = head
          var fast = head.next.next
          while (fast != null && fast.next != null) {
            slow = slow.next
            fast = fast.next.next
          }
          val root = TreeNode(slow.next.x)
          val r = slow.next.next
          slow.next.next = null
          slow.next = null
          root.left = sortedListToBST(head)
          root.right = sortedListToBST(r)
          root
        }
    }
}

File: leetcodeJava/109.java
package leetcodeJava;

class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        if (head == null) return null;
        ListNode slow = head, fast = head.next;
        ListNode prev = null;
        while (fast != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next;
            if (fast != null) fast = fast.next;
        }
        var root = new TreeNode(slow.val);
        if (prev != null) prev.next = null;
        if (slow == head) root.left = null;
        else root.left = sortedListToBST(head);
        root.right = sortedListToBST(slow.next);
        return root;
    }
}
